scripts/linera_net_helper.sh

```diff
@@ -9,32 +9,24 @@ function linera_spawn_and_read_wallet_variables() {
     # Clean up the temporary directory in case we exit.
     trap 'rm -rf "$LINERA_TMP_DIR"' EXIT
 
-    LINERA_TMP_PID="$LINERA_TMP_DIR/pid"
     LINERA_TMP_OUT="$LINERA_TMP_DIR/out"
     LINERA_TMP_ERR="$LINERA_TMP_DIR/err"
     mkfifo "$LINERA_TMP_ERR" || exit 1
 
-    (
-        # Ignoring SIGPIPE to keep `tee` alive after `sed` exits below, closing
-        # LINERA_TMP_ERR.
-        trap '' PIPE
-
-        # Start the main process.
-        "$@" 2> >(tee "$LINERA_TMP_ERR" 2>/dev/null) 1>"$LINERA_TMP_OUT" &
-
-        # Remember the PID of the service for later.
-        jobs -p > "$LINERA_TMP_PID"
-    )
-    LINERA_NET_PID=$(cat "$LINERA_TMP_PID")
+    "$@" >"$LINERA_TMP_OUT" 2>"$LINERA_TMP_ERR" &
+    LINERA_NET_PID=$!
 
     # When the shell exits, we will clean up the top-level jobs (if any), the temporary
     # directory, and the main process. Handling future top-level jobs here is useful
     # because calling `trap` again will be replace this handler.
     trap 'jobs -p | xargs -r kill && rm -rf "$LINERA_TMP_DIR" && kill "$LINERA_NET_PID"' EXIT
 
     # Read from LINERA_TMP_ERR until the string "READY!" is found.
-    sed -n '/^READY!/q' <"$LINERA_TMP_ERR" || exit 1
+    sed '/^READY!/q' <"$LINERA_TMP_ERR" || exit 1
 
     # Source the Bash commands output by the server.
     source "$LINERA_TMP_OUT"
+
+    # Continue to output the command's stderr onto stdout.
+    cat "$LINERA_TMP_ERR" &
```

Comment:
Is it correct to say to the named pipe is still alive here because bash still has a file descriptor on it (created at line 25 to populate the stdin of sed but not yet garbage-collected)? This seems to be the crucial idea that I missed before.

Reply:
The named pipe is alive until you unlink the file name anyway.  `SIGPIPE` happens not when the pipe ‘dies’ but when there are no readers left on the other end of it (i.e. nobody has an open FD pointing to the reading end).  `tee` errors in your original formulation because you're asking it to deliver output to two different sinks, but one of those sinks (possibly temporarily!) has no reader — so it has the choice of:
a) block (or exit) and don't deliver the data to the other sink (stdout) either
b) grow its memory usage indefinitely to buffer all the data, then replay it if a new process connects to the pipe
c) drop the data that it would have delivered, and if a new process connects to the pipe start sending it data from an arbitrary point based on the processing speed of the source and the other sink (bearing in mind that these are just byte streams with no message framing, so an ‘arbitrary point’ quite possibly renders the rest of the stream useless depending on how stateful the protocol is)

`tee` chooses c) as the best of some bad options, but it makes it very clear to you that it's not happy about it.  However, for what the script wants to do, only the first part of the script — until `READY!` — actually needs to be `tee`d.  Instead of having `tee` run the whole time and then close one of its sinks on it when we no longer care about the output, we instead pass the data through the named pipe linearly, and ‘tee’ it on the other side (by allowing `sed` to print its pattern space).  This has the advantage that when we're done tee-ing the output, we can switch to just passing it through linearly (`sed` ends and `cat` begins).  When this happens (`sed` ends) `linera net up` will receive `SIGPIPE` just as `tee` did, but it doesn't do anything with it, and when we continue reading the output it continues normally.  We could avoid the SIGPIPE entirely by opening the named pipe in the bash process itself, which would work as you suggest:

```bash
exec {LINERA_ERR_FD}<"$LINERA_TMP_ERR"
…
sed … <&$LINERA_ERR_FD
cat <&$LINERA_ERR_FD 
```

Unfortunately macOS only comes with bash 3, and the `{FD}<` syntax to choose an unused file descriptor is only supported from 4 onwards (we're on bash 5 now).  Finding an unused FD is otherwise rather inconvenient and would add significant complexity to the script (AFAICT you basically have to loop over `/dev/fd`).

Reply:
Similarly, we can simplify this function somewhat by using local variables instead of global variables with a `LINERA_` prefix, but I'm not sure when local variables were introduced to bash, and I don't have a copy of bash 3 available to test on.

Reply:
I can try in a moment. The PID one couldn't be local but we just removed it.